Only one instance of OpenSBI 0.6

JohnAZoidberg opened the pull request:
OpenSBI can only be initialized once. Therefore we do it as early as possible -
in SEC, while switching to PEI.
To switch to DXE later from PEI we call sbi_hart_switch_mode. This function is saved in the firmware specific context.
Because we don't have access to the OpenSBI library later in PEI we create the PPI to call OpenSBI in SEC.
In DXE SBI is called via the ecall mechanism.

TODO:
- [ ] Implement the PPI functions
- [ ] Remove all usages of RiscVOpensbiLib except in SecMain
  - This is almost done except [here](https://github.com/changab/edk2-platforms/blob/master/Platform/SiFive/U5SeriesPkg/Library/SerialIoLib/SerialPortLib.c). This needs not OpenSBI directly but the `util` part. Maybe we could split this into its own Library? Or add a PPI to use it?
- [ ] Maybe wrap the OpenSBI headers?

###### Hint
See the latest 3 commits, the others are cherry-picks of the previously implement PPI.

JohnAZoidberg commented on RiscVPkg/Library/RiscVDxeIplHandoffOpenSbiLib/RiscVDxeIplHandoffOpenSbiLib.c:
```diff
@@ -105,5 +112,13 @@ RiscVPlatformHandOffToDxeCore (
-  sbi_init (ThisScratch);
+  //sbi_init (ThisScratch);
+  UINTN ThisHartid = sbi_current_hartid();
+  FirmwareContext->HartSpecific[ThisHartid]->HartSwitchMode(
```
Do we need the intermediate `RiscVDxeIplHandoffOpenSbiHandler` function or can we directly call into DXE Core?
Is it necessary to call SwitchStack?

changab replied:
We still need this wrapper function coz sbi_hart_switch_mode only deals with two arguments, however we have three args to pass to DXe core entry point.

JohnAZoidberg commented on RiscVPlatformPkg/Universal/Sec/SecMain.c:
```diff
@@ -458,6 +494,35 @@ static VOID EFIAPI PeiCore(VOID)
+/**
+  RISC-V SEC to PEI handler
+
+  We can't directly transition to PEI because sbi_init() doesn't pass the arguments that the PEI core entry point expects.
```
Actually because we call our own `PeiCore` function and not the real PEI Core directly, we can `sbi_init` to this. `RiscVSecOpenSbiHandler` shouldn't be needed.

changab replied:
seems so, if we dont need HartId information passed from sbi_hart_switch_mode()

JohnAZoidberg replied:
Done

JohnAZoidberg commented on RiscVPlatformPkg/Universal/Sec/SecMain.c:
```diff
@@ -514,10 +580,13 @@ VOID EFIAPI SecCoreStartUpWithStack(UINTN hartid, struct sbi_scratch *scratch)
-    sbi_ecall_init(); // Initial ecall registration in SEC phase for handling further traps.
-    PeiCore();
+  if (HartId == FixedPcdGet32(PcdBootHartId)) {
+    // Doesn't work when I put an argument here :/ With 1 as arg I get an access fault
```
cc @changab

changab replied:
are you sure the exception happens right after "scratch->next_arg1 = (UINTN)1;", or it happens later?
So, actually we can set Scratch->next_addr to PeiCore() in this file, without giving arg1 a value.

JohnAZoidberg replied:
No, it happens somewhere during `sbi_init`.
Yes, we don't need to pass anything else, so I'll just remove this.

JohnAZoidberg committed:
RiscVPlatformPkg: Remove SerialPortLib

It's not used.

JohnAZoidberg committed:
RiscVPkg: Remove usages of RiscVOpensbiLib

changab commented on RiscVPkg/Include/IndustryStandard/RiscVOpensbi.h:
```diff
@@ -102,14 +102,26 @@
+    RISCV_UINT128          MachineVendorId;        // Machine vendor ID
+    RISCV_UINT128          MachineArchId;          // Machine Architecture ID
+    RISCV_UINT128          MachineImplId;          // Machine Implementation ID
+    RISCV_HART_SWITCH_MODE HartSwitchMode;
```
It would be better to create another structure for internal functions if consider more functions would be public.

JohnAZoidberg replied:
Good idea

JohnAZoidberg replied:
I can't quite figure out how to do that. I'm not sure where to put this structure in `RiscVOpenSbiHartSwitchMode` and why it needs to be 64 byte aligned.

changab replied:
Just something like this

typedef struct {
    UINT64          IsaExtensionSupported;  // The ISA extension this core supported.
    RISCV_UINT128   MachineVendorId;        // Machine vendor ID
    RISCV_UINT128   MachineArchId;          // Machine Architecture ID
    RISCV_UINT128   MachineImplId;          // Machine Implementation ID
   **EDK2_OPEN_SBI_INTERNAL_FUNCTION OpensbiInternalFunc;**
} EFI_RISCV_FIRMWARE_CONTEXT_HART_SPECIFIC;

typedef struct {
    RISCV_HART_SWITCH_MODE HartSwitchMode;
   ...
   ...
}    EDK2_OPEN_SBI_INTERNAL_FUNCTION


64-bit alignment for load/store using 64-bit regs is the performance consideration. and some hardware may not support misaligned access.

changab commented on RiscVPlatformPkg/Universal/Sec/RiscVOpenSbiPpi.c:
```diff
@@ -0,0 +1,343 @@
+#include <sbi/sbi_types.h>
+#include <sbi/sbi_platform.h>
+#include <sbi/sbi_init.h>
+
```
we need function header for each function.

JohnAZoidberg replied:
Yup, I already wrote them and put them in the header. I'll check them and copy them here.

changab commented on RiscVPlatformPkg/Universal/Sec/SecMain.c:
```diff
@@ -458,6 +493,35 @@ static VOID EFIAPI PeiCore(VOID)
+EFIAPI
+RiscVOpenSbiHartSwitchMode (
+  IN  UINTN   Arg0,
+  IN  UINTN   Arg1,
```
Would you like to name these variables more understandable? such as Arg0 -> FucntionArg0  or something

JohnAZoidberg replied:
Okay

JohnAZoidberg committed:
RiscVPkg: Bump OpenSBI to 0.6

JohnAZoidberg committed:
RiscV: Only init OpenSBI once, in SEC

OpenSBI can only be initialized once. Therefore we do it as early as possible -
in SEC, while switching to PEI.
To switch to DXE later from PEI we call sbi_hart_switch_mode. This function is saved in the firmware specific context.
Because we don't have access to the OpenSBI library later in PEI and DXE we create the PPI to call OpenSBI in SEC.

JohnAZoidberg committed:
WIP: RiscVPlatforms: Install OpenSBI PPIs in SEC

SEC installs PPIs for later stages to call SBI functions.
OpenSBI cannot be initialize multiple times, therefore SEC does it.

Each SBI extension has its own PPI:
- gPeiOpenSbiBasePpiGuid
- gPeiOpenSbiLegacyPpiGuid
- gPeiOpenSbiHsmPpiGuid
- gPeiOpenSbiVendorPpiGuid
- gPeiOpenSbiLibraryPpiGuid

TODO:
- Implement calling OpenSBI syscall handler from PPI
- Add PPIs for new extensions in SBI 0.2

JohnAZoidberg commented:
@changab I addressed your comments and updated OpenSBI to 0.6

Please:
- cherry-pick/squash d2c45a3c244216439681cde5de36d3e97e813c4c onto the RISC-V-V2 branch.
- ignore dbff4a924fc113133cff1463987598d60266839d. We don't need the PPI right now.
- merge the rest.

JohnAZoidberg commented:
changab/edk2-platforms#3 is related.

changab commented:
looks neat now.

JohnAZoidberg commented:
Still boots to properly working UEFI Shell btw.
